### PR TITLE
VTK volume viewer export color mapper informations to python side

### PR DESCRIPTION
**Pull request by xavArtley**
With this PR we begin to have a real 3d slicer
```python
import panel as pn
pn.extension('vtk')
import holoviews as hv
hv.extension('bokeh')
import pyvista as pv
from pyvista import examples
# Download a volumetric dataset
head = examples.download_head()
arr = head.active_scalar.reshape(head.dimensions, order='F')
interpolation_level = pn.widgets.IntSlider(value=5, start=1, end=20, name='Interpolation Level')
vol = pn.panel(head, sizing_mode='stretch_both', height=400, display_slices=True, rescale=True, display_volume=False,
              orientation_widget=True)
import scipy.ndimage as sn

@pn.depends(vol.param.slice_i,vol.param.slice_j,vol.param.slice_k,vol.param.mapper, interpolation_level.param.value)
def print_mapper(si, sj, sk, mapper, interp_level):
    im_interp = lambda x: sn.zoom(x, zoom=interp_level, order=1)
    low = mapper['low'] if mapper else arr.min()
    high = mapper['high'] if mapper else arr.max()
    cmap = mapper['palette'] if mapper else 'fire'
    
    im_i = hv.Image(im_interp(arr[si,::-1,:]), bounds=(head.extent[4],head.extent[2],head.extent[5],head.extent[3]), kdims=['z','y'], vdims='Intensity')
    im_j = hv.Image(im_interp(arr[::-1,sj,:]), bounds=(head.extent[4],head.extent[0],head.extent[5],head.extent[1]), kdims=['z','x'], vdims='Intensity')
    im_k = hv.Image(im_interp(arr[::-1,:,sk]), bounds=(head.extent[2],head.extent[0],head.extent[3],head.extent[1]), kdims=['y','x'], vdims='Intensity')
    return (im_i + im_j + im_k).opts(hv.opts.Image(clim=(low, high), cmap=cmap, invert_axes=True, data_aspect=1))
    
    
pn.Column(
    pn.Row(
        pn.Column(vol.controls(jslink=False,
                               parameters=['render_background', 'rescale', 'slice_i', 'slice_j', 
                                           'slice_k', 'display_volume', 'display_slices']),
                  interpolation_level),
        vol
    ),
    print_mapper
)
```
ezgif com-video-to-gif (1)

### CODE REVIEW

**Comment by philippjfr** [comment]
Ready to merge?

**Comment by xavArtley** [comment]
No I'd like to modify documentation before

**Comment by xavArtley** [comment]
What do you think to had something like that to the gallery?
```python
import param
import panel as pn
import holoviews as hv
import pyvista as pv
from pyvista import examples
from scipy.ndimage import zoom

pn.extension('vtk')
hv.extension('bokeh')


class ImageSmoother(param.Parameterized):
    
    smooth_fun = param.Parameter(default=None)
    order = param.Selector(default=1, objects=[1,2,3])
    zoom = param.Integer(default=2, bounds=(1,10))
    
    def __init__(self, **params):
        super().__init__(**params)
        self._update_fun()
    
    @param.depends('order', 'zoom', watch=True)
    def _update_fun(self):
        self.smooth_fun = lambda x: zoom(x, zoom=self.zoom, order=self.order)
    
    @property
    def stream(self):
        return hv.streams.Params(self, parameters=['smooth_fun'])



# Download a volumetric dataset
vol = examples.download_head()
volume = pn.panel(vol, sizing_mode='stretch_both', height=400, display_slices=True, orientation_widget=True)
volume_controls = volume.controls(jslink=False, parameters=['render_background', 'display_volume', 'display_slices',
                                                            'slice_i', 'slice_j', 'slice_k', 'rescale'])
toggle_parallel_proj = pn.widgets.Toggle(name='Parallel Projection', value=False)
def update_camera_projection(*evts):
    volume.camera['parallelProjection'] = evts[0].new
    volume.param.trigger('camera')
toggle_parallel_proj.param.watch(update_camera_projection, ['value'], onlychanged=True)



stream_i = hv.streams.Params(volume, parameters=['slice_i'], rename={'slice_i':'si'})
stream_j = hv.streams.Params(volume, parameters=['slice_j'], rename={'slice_j':'sj'})
stream_k = hv.streams.Params(volume, parameters=['slice_k'], rename={'slice_k':'sk'})
stream_mapper = hv.streams.Params(volume, parameters=['mapper'])
smoother = ImageSmoother()
stream_smooth = smoother.stream

arr = vol.active_scalar.reshape(vol.dimensions, order='F')

def image_slice_i(si, mapper, smooth_fun):
    low = mapper['low'] if mapper else arr.min()
    high = mapper['high'] if mapper else arr.max()
    cmap = mapper['palette'] if mapper else 'fire'
    lbrt = vol.extent[4], vol.extent[2], vol.extent[5], vol.extent[3]
    im_i = hv.Image(smooth_fun(arr[si,::-1,:]), bounds=lbrt, kdims=['z','y'], vdims='Intensity')
    frame_width = 500
    frame_height = int(frame_width * (lbrt[2]-lbrt[0])/(lbrt[3]-lbrt[1]))
    return im_i.opts(hv.opts.Image(clim=(low, high), cmap=cmap, frame_width=frame_width, frame_height=frame_height))
    
def image_slice_j(sj, mapper, smooth_fun):
    low = mapper['low'] if mapper else arr.min()
    high = mapper['high'] if mapper else arr.max()
    cmap = mapper['palette'] if mapper else 'fire'
    lbrt = vol.extent[4], vol.extent[0], vol.extent[5], vol.extent[1]
    im_j = hv.Image(smooth_fun(arr[::-1,sj,:]), bounds=lbrt, kdims=['z','x'], vdims='Intensity')
    frame_width = 500
    frame_height = int(frame_width * (lbrt[2]-lbrt[0])/(lbrt[3]-lbrt[1]))
    return im_j.opts(hv.opts.Image(clim=(low, high), cmap=cmap, frame_width=frame_width, frame_height=frame_height))
                     
def image_slice_k(sk, mapper, smooth_fun):
    low = mapper['low'] if mapper else arr.min()
    high = mapper['high'] if mapper else arr.max()
    cmap = mapper['palette'] if mapper else 'fire'
    lbrt = vol.extent[2], vol.extent[0], vol.extent[3], vol.extent[1]
    im_k = hv.Image(smooth_fun(arr[::-1,:,sk]), bounds=lbrt, kdims=['y','x'], vdims='Intensity')
    frame_width = 500
    frame_height = int(frame_width * (lbrt[2]-lbrt[0])/(lbrt[3]-lbrt[1]))
    return im_k.opts(hv.opts.Image(clim=(low, high), cmap=cmap, frame_width=frame_width, frame_height=frame_height))

hv.opts.defaults(hv.opts.Image(invert_axes=True, data_aspect=1))
dmap_i = hv.DynamicMap(image_slice_i, streams=[stream_i, stream_mapper, stream_smooth])
dmap_j = hv.DynamicMap(image_slice_j, streams=[stream_j, stream_mapper, stream_smooth])
dmap_k = hv.DynamicMap(image_slice_k, streams=[stream_k, stream_mapper, stream_smooth])


controller = pn.WidgetBox(toggle_parallel_proj, *volume_controls[1:], pn.Param(smoother, parameters=['zoom', 'order']))
pn.Row(pn.Column(pn.Row(controller, volume), pn.Row(dmap_i, dmap_j)), dmap_k, sizing_mode='stretch_height').show()
```

**Comment by xavArtley** [comment]
https://files.gitter.im/pyviz/pyviz/UEw2/ezgif.com-video-to-gif-2-.gif

**Comment by philippjfr** [comment]
Wow, how did you do the draggable layout?

**Comment by xavArtley** [comment]
I used http://golden-layout.com/

**Comment by philippjfr** [comment]
Very cool, you used it as a custom template?

**Comment by xavArtley** [comment]
Yes i tried to mix different aspects of panel

Le jeu. 6 févr. 2020 à 18:27, Philipp Rudiger <notifications@github.com> a
écrit :

File: panel/models/vtk/vtk_layout.ts [hunk 1]
```diff
@@ -1,14 +1,136 @@
...
+  protected _widgetManager: any
+  protected _setting_camera: boolean = false
+
+  connect_signals(): void {
+    super.connect_signals()
+    this.connect(this.model.properties.data.change, () => {
+      this.invalidate_render()
+    })
+    this.connect(this.model.properties.orientation_widget.change, () => {
+      this._orientation_widget_visbility(this.model.orientation_widget)
```
Comment (philippjfr) [inline panel/models/vtk/vtk_layout.ts]: visibility is misspelled, but at least consistently so it won't hold up merge.

**Comment by philippjfr** [comment]
Going to merge and get this in 0.8.1 pretending it's a simple enhancement :smile:

File: panel/models/vtk/vtk_layout.ts (hunk 1 above)
Comment (xavArtley) [reply panel/models/vtk/vtk_layout.ts]: Indeed I pretty good to misspelled it

**Comment by philippjfr** [comment]
@xavArtley The VTKVolume.mapper isn't syncing for me in the notebook, any idea what's going on there?

**Comment by xavArtley** [comment]
Actually for me it's not sync too in the notebook but only when I use the
tempate. If I use panel Row and Columns it works as expected.

Le sam. 8 févr. 2020 à 00:51, Philipp Rudiger <notifications@github.com> a
écrit :

File: examples/gallery/demos/VTKSlicer.ipynb [hunk 2]
```diff
@@ -250,16 +267,16 @@
...
+    "\n",
+    "tmpl.add_panel('controller', controller)\n",
     "tmpl.add_panel('scene3d', volume)\n",
     "tmpl.add_panel('slice_i', pn.panel(dmap_i, sizing_mode='stretch_both'))\n",
     "tmpl.add_panel('slice_j', pn.panel(dmap_j, sizing_mode='stretch_both'))\n",
     "tmpl.add_panel('slice_k', pn.panel(dmap_k, sizing_mode='stretch_both'))\n",
     "\n",
-    "\n",
-    "tmpl.servable(title='VTKSlicer').show()"
+    "tmpl.servable(title='VTKSlicer').servable()"
```
Comment (xavArtley) [inline examples/gallery/demos/VTKSlicer.ipynb]: 2 times call to servable?

**Comment by xavArtley** [comment]
I haven't read all the template documentation to handle template in notebook. I will make some modifications instead of testing the page title

**Comment by xavArtley** [comment]
> @xavArtley The VTKVolume.mapper isn't syncing for me in the notebook, any idea what's going on there?

I found something strange when looking at `volume.model` after displaying the template there is 2 reférences
Capture d’écran de 2020-02-08 13-16-08

**Comment by philippjfr** [comment]
>I found something strange when looking at volume.model after displaying the template there is 2 reférences

This is in fact expected, templates have to do some funny stuff to linking set up.

File: examples/gallery/demos/VTKSlicer.ipynb (hunk 2 above)
Comment (philippjfr) [reply examples/gallery/demos/VTKSlicer.ipynb]: Oops accident.

**Comment by xavArtley** [comment]
@philippjfr Do you think it should be possible to set data_aspect=1 while keeping layout responsivness?

**Comment by philippjfr** [comment]
>@philippjfr Do you think it should be possible to set data_aspect=1 to images while keeping layout responsivness?

Would definitely need some special handling I think only bokeh figures currently support aspect handling out of the box.

**Comment by xavArtley** [comment]
@philippjfr Python 2 failed on master because the property `.active_scalars` was introduced on pyvista 0.23.1 and for the python 2 build the version 0.22.4 seems to be installed.
2 solutions to correct it:
 - use the old property (trigger deprection warnings) `.active_scalar`
 - update pyvista version to >= 0.23for python 2 build however I don't know why it's not the same version on python 3 and python 2.

**Comment by philippjfr** [comment]
@xavArtley I'm happy to defer to your judgement here. Given that we are dropping python 2 support in about a week and there will be only one more release, I'm happy to skip this test for py2 if pyvista>=0.23 doesn't support py2.

**Comment by JeffreyWardman** [comment]
The notebook for this (VTKSlicer.ipynb) does not successfully link the controls with the image. E.g. slicing does not work, neither does selecting the background.

**Comment by xavArtley** [comment]
Are you talking about it in the panel documentation or when you run the notebook locally?
It is not working in the documentation because it needs a server
however locally I tested it and it seems ok:
vtkslicer

**Comment by JeffreyWardman** [comment]
@xavArtley I was testing locally. It had to do with the Jupyter extensions. It didn't return anything in Jupyter notebook. In Jupyter lab it showed the window but didn't have any interactivity. Turns out the reason was because I hadn't installed the pyviz extension. Problem solved! Thanks for your help.

**Pull request merged**